rust/numaflow-core/src/reduce/reducer/unaligned/windower/session.rs

```diff
@@ -118,27 +116,41 @@ impl SessionWindowManager {
         }]
     }
 
-    /// expands the window if the start or end time collides.
+    /// expands the window if the windows overlap.
+    /// If windows overlap, always expand to accommodate both windows.
+    /// If the new window is completely contained within existing boundaries, just append to existing window.
+    /// Windows must overlap first (considering end time is exclusive) before they can be expanded.
     fn expand_window_if_needed(new_window: &Window, existing_window: &Window) -> Option<Window> {
-        (new_window.start_time < existing_window.start_time
-            || new_window.end_time > existing_window.end_time)
-            .then(|| {
-                Window::new(
-                    new_window.start_time.min(existing_window.start_time),
-                    new_window.end_time.max(existing_window.end_time),
-                    Arc::clone(&new_window.keys),
-                )
-            })
+        // First check if windows overlap (end time is exclusive, start time is inclusive)
+        // [a,b) and [c,d) overlap if a < d and c < b
```

Comment:
```suggestion
        // [a,b) and [c,d) overlap if a < c and c < b
```

let's c < d, this is given